indra/newview/lluilistener.cpp

```diff
@@ -97,6 +99,41 @@ LLUIListener::LLUIListener():
         "Add menu separator to the [\"parent_menu\"] within the Top menu.",
         &LLUIListener::addMenuSeparator,
         llsd::map("parent_menu", LLSD(), "reply", LLSD()));
+
+    add("setMenuVisible",
+        "Set menu [\"name\"] visibility to [\"visible\"]",
+        &LLUIListener::setMenuVisible,
+        llsd::map("name", LLSD(), "visible", LLSD(), "reply", LLSD()));
+
+    add("defaultToolbars",
+        "todo: defaultToolbars desc",
+        &LLUIListener::restoreDefaultToolbars);
+
+    add("clearToolbars",
+        "Clear all buttons off the toolbars",
+        &LLUIListener::clearAllToolbars);
+
+    add("addToolbarBtn",
+        "Add [\"btn_name\"] toolbar button to the [\"toolbar\"]\n"
+        "[1 (left toolbar), 2 (right toolbar), 3 (bottom toolbar)]\n"
+        "Position of the command in the original list can be specified as [\"rank\"]",
+        &LLUIListener::addToolbarBtn,
+        llsd::map("btn_name", LLSD(), "reply", LLSD()));
```

Comment:
How are `"btn_name"` buttons normally defined? Can we add a `"btn_name"` that invokes a Lua script?

Since `"toolbar"` is optional, mention that the default is bottom -- that's not at all obvious.

Wouldn't it be better to accept a string `"left"`, `"right"`, `"bottom"`?

It's worth mentioning (especially for Lua callers) that `rank` starts at 0.

Reply:
Toolbar buttons are defined in app_settings\commands.xml. It depends. Adding just one button and define which script it invokes - trivial. Ability to add multiple new buttons that will not persist between sessions and have custom callbacks - may be tricky. (I need to think about it).
Agree. Using "left", "right", "bottom" seems to be better.

Reply:
Problem of scale?